creusot/src/backend/dependency.rs

```diff
@@ -79,25 +40,62 @@ impl<'tcx> Dependency<'tcx> {
     ) -> Option<Self> {
         match self {
             Dependency::Type(ty) => resolve_type(ty, ctx.tcx, param_env),
-            Dependency::Item(Key { did: item, subst: substs }) => {
-                resolve_item(item, substs, ctx.tcx, param_env)
-            }
+            Dependency::Item(item, substs) => resolve_item(item, substs, ctx.tcx, param_env),
+            dep @ Dependency::TyInv(_) => Some(dep),
         }
     }
 
-    pub(crate) fn cloneable_id(self) -> Option<Key<'tcx>> {
+    pub(crate) fn did(self) -> Option<(DefId, SubstsRef<'tcx>)> {
         match self {
-            Dependency::Item(i) => Some(i),
-            Dependency::Type(t) => match t.kind() {
-                TyKind::Adt(def, substs) => Some(Key { did: def.did(), subst: substs }),
-                TyKind::Closure(id, substs) => Some(Key { did: *id, subst: substs }),
-                TyKind::Alias(AliasKind::Projection, aty) => {
-                    Some(Key { did: aty.def_id, subst: aty.substs })
-                }
+            Dependency::Item(def_id, subst) => Some((def_id, subst)),
+            Dependency::Type(t) | Dependency::TyInv(t) => match t.kind() {
+                TyKind::Adt(def, substs) => Some((def.did(), substs)),
+                TyKind::Closure(id, substs) => Some((*id, substs)),
+                TyKind::Alias(AliasKind::Projection, aty) => Some((aty.def_id, aty.substs)),
                 _ => None,
             },
         }
     }
+
+    pub(crate) fn is_inv(&self) -> bool {
+        matches!(self, Dependency::TyInv(_))
+    }
+
+    #[inline]
+    pub(crate) fn erase_regions(mut self, tcx: TyCtxt<'tcx>) -> Self {
+        match &mut self {
+            Dependency::Item(_, s) => *s = tcx.erase_regions(*s),
```

Comment:
I wonder if you couldn't derive `TypeFoldable` for `Dependency` and get this for free?